sequence_processing_pipeline/Pipeline.py

```diff
@@ -292,14 +317,220 @@ def generate_sample_information_files(self):
         return paths
 
     def get_sample_ids(self):
-        return [x.Sample_ID for x in self.sample_sheet.samples]
+        # test for self.mapping_file, since self.sample_sheet will be
+        # defined in both cases.
+        if self.mapping_file is not None:
+            results = []
+            sample_project_map = {}
+            for sample, project in zip(self.mapping_file.sample_name,
+                                       self.mapping_file.project_name):
+                if project not in sample_project_map:
+                    sample_project_map[project] = []
+                sample_project_map[project].append(sample)
+
+            for project in sample_project_map:
+                results += sample_project_map[project]
+        else:
+            results = [x.Sample_ID for x in self.sample_sheet.samples]
+
+        return results
 
     def get_project_info(self):
-        bioinformatics = self.sample_sheet.Bioinformatics
+        # test for self.mapping_file, since self.sample_sheet will be
+        # defined in both cases.
         results = []
 
-        for result in bioinformatics.to_dict('records'):
-            results.append({'project_name': result['Sample_Project'],
-                            'qiita_id': result['QiitaID']})
+        if self.mapping_file is not None:
+            sample_project_map = {}
+            for sample, project in zip(self.mapping_file.sample_name,
+                                       self.mapping_file.project_name):
+                if project not in sample_project_map:
+                    sample_project_map[project] = []
+                sample_project_map[project].append(sample)
+
+            # TODO: Replace 'something' w/proper Qiita ID
+            for project in sample_project_map:
+                qiita_id = project.split('_')[-1]
+                results.append(
+                    # assume project names end in qiita ids
+                    {'project_name': project, 'qiita_id': qiita_id})
+        else:
+            bioinformatics = self.sample_sheet.Bioinformatics
+            for result in bioinformatics.to_dict('records'):
+                results.append({'project_name': result['Sample_Project'],
+                                'qiita_id': result['QiitaID']})
 
         return results
+
+    @staticmethod
+    def is_mapping_file(mapping_file_path):
+        try:
+            Pipeline._validate_mapping_file(mapping_file_path)
+            return True
+        except PipelineError:
+            return False
+
+    @staticmethod
+    def _validate_mapping_file(mapping_file_path):
+        exp = {'barcode', 'library_construction_protocol', 'mastermix_lot',
```

Comment:
My concern with this is that we will be growing these headers in the future ... will this break with new headers?

Reply:
It is true that it will break if columns are added or subtracted. We could change the logic so that as long as the columns in exp exist, then any additional columns don't break validation.

Reply:
Yes, let's change the logic.

Reply:
My mistake - the logic will work correctly as long as the columns in the mapping file are equal or superset of exp. It only returns an error if the mapping file is missing a column.